File: Python/Medium/2807.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
def gcd(a: int, b: int) -> int:
    while b:
        a, b = b, a % b
    return a


class Solution:
    def insertGreatestCommonDivisors(
        self, head: Optional[ListNode]
    ) -> Optional[ListNode]:

        temp = head

        while temp.next:
            value = gcd(temp.val, temp.next.val)

            holder = ListNode(val=value, next=temp.next)
            temp.next = holder

            # go to the new node
            temp = temp.next
            # go to the actual next node
            temp = temp.next

        return head

File: Java/Medium/2807.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode insertGreatestCommonDivisors(ListNode head) {
        ListNode temp = head;

        while (temp.next != null) {
            int first = temp.val;
            int second = temp.next.val;

            int value = gcd(first, second);

            ListNode holder = new ListNode(value, temp.next);
            temp.next = holder;

            // go to the newly added node
            temp = temp.next;
            // actual next node
            temp = temp.next;
        }

        return head;

    }

    public int gcd(int first, int second) {

        int answer = (first > second) ? second : first;
        while (answer > 1) {
            if (first % answer == 0 && second % answer == 0) {
                break;
            }

            answer--;
        }

        return answer;
    }
}